server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java

```diff
@@ -0,0 +1,220 @@
+                    var changesResponse = getChanges(fromSeqNo, toSeqNo);
+                    LOGGER.debug("[{}] Got {} changes starting from seqNo: {}",
+                                 shardId, changesResponse.changes().size(), fromSeqNo);
+                    sequencer.send(changesResponse);
+                    LOGGER.debug("[{}] Pushed to sequencer {}-{}", shardId, fromSeqNo, toSeqNo);
+                    long lastSeqNo = fromSeqNo - 1;
+                    if (changesResponse.changes().isEmpty() == false) {
+                        lastSeqNo = changesResponse.changes().get(changesResponse.changes().size() - 1).seqNo();
+                    }
+                    updateBatchFetched(
+                        true,
+                        fromSeqNo,
+                        toSeqNo,
+                        lastSeqNo,
+                        changesResponse.lastSyncedGlobalCheckpoint()
+                    );
+                } catch (ElasticsearchTimeoutException e) {
+                    LOGGER.debug("[{}] Timed out waiting for new changes. Current seqNo: {}", shardId, fromSeqNo);
+                    updateBatchFetched(false, fromSeqNo, toSeqNo, fromSeqNo - 1, -1);
+                } catch (NodeNotConnectedException e) {
+                    LOGGER.info("[{}] Node not connected. Retrying request using a different node. {}",
+                                shardId, e.getStackTrace());
+                    // TODO: retry using different node
+                    //delay(backOffForNodeDiscovery)
+                    updateBatchFetched(false, fromSeqNo, toSeqNo, fromSeqNo - 1, -1);
+                } catch (Exception e) {
+                    LOGGER.error("[{}] Unable to get changes from seqNo: {}. {}",
+                                 shardId, fromSeqNo, e.getStackTrace());
+                    updateBatchFetched(false, fromSeqNo, toSeqNo, fromSeqNo - 1, -1);
+                }
+
+            })
+        );
+    }
+
+    /**
+     * Provides a range of operations to be fetched next.
+     */
+    public void requestBatchToFetch(Consumer<Tuple<Long, Long>> consumer) {
+        LOGGER.debug("[{}] Waiting to get batch. requested: {}, leader: {}",
+                     shardId, seqNoAlreadyRequested.get(), observedSeqNoAtLeader.get());
+
+        // Wait till we have batch to fetch. Note that if seqNoAlreadyRequested is equal to observedSeqNoAtLeader,
+        // we still should be sending one more request to fetch which will just do a poll and eventually timeout
+        // if no new operations are there on the leader (configured via TransportGetChangesAction.WAIT_FOR_NEW_OPS_TIMEOUT)
+        if (seqNoAlreadyRequested.get() > observedSeqNoAtLeader.get() && missingBatches.isEmpty()) {
+            cancellable = threadPool.schedule(
+                () -> requestBatchToFetch(consumer),
+                replicationSettings.pollDelay(),
+                ThreadPool.Names.LOGICAL_REPLICATION
+            );
+            return;
+        }
```

Comment:
Based on the class name and the javadoc I'd assume that the `ShardReplicationChangesTracker` operates indefinitely, but this looks as if it will stop once it has caught up with all changes. How will replication continue once the shards in the publication advance further?